### DISCUSSION TITLE
riverpod_lint: I don't understand one of the lints

### DISCUSSION THREAD

**Discussion by julienlebren** [rrousselGit/riverpod#2238] [Q&A] [answered]
This code is a common code to many of my apps, it creates a `PurchasesService`.
Each app sets its own  `PurchasesSettings` by overriding the provider in the very first `ProviderScope` of the app.

```dart
@Riverpod(keepAlive: true)
PurchasesSettings purchasesSettings(PurchasesSettingsRef ref) {
  throw UnimplementedError(
    "You need to override purchasesSettingsProvider before calling purchasesServiceProvider!",
  );
}

@Riverpod(keepAlive: true, dependencies: [purchasesSettings])
PurchasesService purchasesService(PurchasesServiceRef ref) {
  final settings = ref.watch(purchasesSettingsProvider);
  return PurchasesService(settings: settings);
}
```

The _purchasesService_ line returns the following lint:
`If a provider depends on providers which specify "dependencies", they should themselves specify "dependencies" and include all the scoped providers.`

I don't understand what I am supposed to do, especially since everything works as expected.

The quick fix only offers to remove the dependencies parameter.

**Comment by rrousselGit** [comment, marked as answer]
`purchasesSettings` doesn't specify `dependencies` and is therefore not "scoped".

Are you overriding it on the root ProviderScope? If so, you don't need to specify dependencies at all.

**Reply by rrousselGit** [reply]
If you do:

```dart
@Riverpod(keepAlive: true, dependencies: [])
PurchasesSettings purchasesSettings(PurchasesSettingsRef ref) {
  throw UnimplementedError(
    "You need to override purchasesSettingsProvider before calling purchasesServiceProvider!",
  );
}
```

Then the lint will disappear

But that means your provider is scoped.

**Comment by julienlebren** [comment]
> Are you overriding it on the root ProviderScope? If so, you don't need to specify dependencies at all.

Ah yes, I was overriding in a "second-level" ProviderScope. I moved everything in the root ProviderScope so I can remove 

> `purchasesSettings` doesn't specify `dependencies` and is therefore not "scoped".

> If you do:
> 
> ```dart
> @Riverpod(keepAlive: true, dependencies: [])
> PurchasesSettings purchasesSettings(PurchasesSettingsRef ref) {
>   throw UnimplementedError(
>     "You need to override purchasesSettingsProvider before calling purchasesServiceProvider!",
>   );
> }
> ```
> 
> Then the lint will disappear
> 
> But that means your provider is scoped.

I still don't understand how I would figure this out (if needed) :
`purchaseService` depends on `purchaseSettings`, that's why I put a dependency on `purchaseService`. 
Why should I put an empty `dependencies` on `purchaseSettings` to avoid the lint? 🤔

**Reply by rrousselGit** [reply]
Adding `dependencies` on `purchasesSettings` is likely not the solution in your case.
Chances are the quick fix was correct and you don't need `dependencies` at all

If you needed `dependencies`, your codebase would have a separate warning for `purchasesSettings`:  [scoped_providers_should_specify_dependencies](https://github.com/rrousselGit/riverpod/tree/master/packages/riverpod_lint#scoped_providers_should_specify_dependencies-generator-only)

**Reply by rrousselGit** [reply]
That other lint is how you'd end-up specifying an empty `dependencies` on your provider

**Reply by julienlebren** [reply]
Does riverpod_lint analyzes only the providers or also the way they are used?
I ask this because the code I posted is in the package, so there is no ProviderScope in the package, but only in the app which is using it, so I wonder how riverpod_lint can say if such provider needs dependencies or not.

**Reply by rrousselGit** [reply]
It analyzes both.

If you use riverpod_lint on your app, then it'll inspect ProviderScopes and their overrides, and check that you're not scoping a provider with no dependencies.

**Comment by julienlebren** [comment]
I have also a problem to understand `scoped_providers_should_specify_dependencies`.
I try to provide a simple example.

```dart
class AppTheme {
  const AppTheme({
    required this.backgroundColor,
    required this.foregroundColor,
  });

  final Color backgroundColor;
  final Color foregroundColor;
}

@riverpod
AppTheme appTheme(AppThemeRef ref) {
  throw UnimplementedError("AppTheme has not been overridden as required.");
}

void main() async {
    runApp(ProviderScope(
      overrides: [
        appThemeProvider.overrideWith(customAppTheme),
      ],
      child: const MyApp(),
    ));
}
```

Then, somewhere in the app where we want to display other colors than the ones defined in the main theme:

```dart
class Widget extends ConsumerWidget {
  const Widget({ Key? key }) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final appTheme = ref.watch(appThemeProvider);

    return ProviderScope(
      overrides: [
        appThemeProvider.overrideWithValue(
          appTheme.copyWith(
            backgroundColor: const Colors.red,
          ),
        ),
      ],
      child: ...
```

This widget displays the `scoped_providers_should_specify_dependencies` lint.
But why? There is no dependency to specify, it depends on nothing 🤷🏻‍♂️

**Reply by rrousselGit** [reply]
You scoped the provider. Scoped providers should specify `dependencies`, even if empty.
This lets `provider_dependencies` know when a provider uses a scoped provider and therefore should include it in `dependencies`

Keep in mind: Only scoped providers need to be specified in `dependencies` now. No need to list **all** dependencies, only the scoped ones.

So a convention was made:
```dart
// not a scoped provider
@riverpod

// this is a scoped provider
@Riverpod(dependencies: [])
```

**Reply by rrousselGit** [reply]
See also the docs about `dependencies` https://pub.dev/documentation/riverpod/latest/riverpod/AutoDisposeStateNotifierProviderFamily/dependencies.html

**Reply by rrousselGit** [reply]
Also note that there's a shorthand for this available now.

Instead of:

```dart
@Riverpod(dependencies: [])
AppTheme appTheme(AppThemeRef ref) {
  throw UnimplementedError("AppTheme has not been overridden as required.");
}
```

You can directly write:

```dart
@riverpod
external AppTheme appTheme();
```

This will both flag the provider as scoped (as if you specified `dependencies: []`) and have the same behaviour (throw by default, can be overridden)

**Reply by julienlebren** [reply]
Ok I got it, especially with `Specifying this list is strictly equivalent to saying "This provider may be scoped".`
But the word 'dependency' is not really correct in such case. Why not create a `canBeScoped` or something like this in the @Riverpod declaratiion?

**Reply by rrousselGit** [reply]
Because that's redundant

`dependencies` is used exclusively for scoped providers, and all scoped providers should specify it.  
Asking folks to instead write `@Riverpod(canBeScoped: true, dependencies: [foo])` isn't an improvement IMO.

**Reply by rrousselGit** [reply]
Also the new shorthand for scoped providers makes this even more redundant.

**Reply by julienlebren** [reply]
Everything is clear now, thanks 👌
